core/gather/gatherers/bf-cache-failures.js

```diff
@@ -0,0 +1,158 @@
+      bfCacheErrorsMap[err.reason] = [];
+    }
+
+    return {notRestoredReasonsTree};
+  }
+
+  /**
+   * @param {LH.Crdp.Page.BackForwardCacheNotRestoredExplanationTree} errorTree
+   * @return {LH.Artifacts.BFCacheFailure}
+   */
+  static processBFCacheEventTree(errorTree) {
+    /** @type {LH.Artifacts.BFCacheNotRestoredReasonsTree} */
+    const notRestoredReasonsTree = {
+      Circumstantial: {},
+      PageSupportNeeded: {},
+      SupportPending: {},
+    };
+
+    /**
+     * @param {LH.Crdp.Page.BackForwardCacheNotRestoredExplanationTree} node
+     */
+    function traverse(node) {
+      for (const error of node.explanations) {
+        const bfCacheErrorsMap = notRestoredReasonsTree[error.type];
+        const frameUrls = bfCacheErrorsMap[error.reason] || [];
+        frameUrls.push(node.url);
+        bfCacheErrorsMap[error.reason] = frameUrls;
+      }
+
+      for (const child of node.children) {
+        traverse(child);
+      }
+    }
+
+    traverse(errorTree);
+
+    return {notRestoredReasonsTree};
+  }
+
+  /**
+   * @param {LH.Crdp.Page.BackForwardCacheNotUsedEvent|undefined} event
+   * @return {LH.Artifacts.BFCacheFailure}
+   */
+  static processBFCacheEvent(event) {
+    if (event?.notRestoredExplanationsTree) {
+      return BFCacheFailures.processBFCacheEventTree(event.notRestoredExplanationsTree);
+    }
+    return BFCacheFailures.processBFCacheEventList(event?.notRestoredExplanations || []);
+  }
+
+  /**
+   * @param {LH.Gatherer.FRTransitionalContext} context
+   * @return {Promise<LH.Crdp.Page.BackForwardCacheNotUsedEvent|undefined>}
+   */
+  async activelyCollectBFCacheEvent(context) {
+    const session = context.driver.defaultSession;
+
+    /** @type {LH.Crdp.Page.BackForwardCacheNotUsedEvent|undefined} */
+    let bfCacheEvent = undefined;
+
+    /**
+     * @param {LH.Crdp.Page.BackForwardCacheNotUsedEvent} event
+     */
+    function onBfCacheNotUsed(event) {
+      bfCacheEvent = event;
+    }
+
+    session.on('Page.backForwardCacheNotUsed', onBfCacheNotUsed);
+
+    const history = await session.sendCommand('Page.getNavigationHistory');
+    const entry = history.entries[history.currentIndex];
+
+    await Promise.all([
+      session.sendCommand('Page.navigate', {url: 'chrome://terms'}),
+      waitForLoadEvent(session, 0).promise,
+    ]);
+
+    await Promise.all([
+      session.sendCommand('Page.navigateToHistoryEntry', {entryId: entry.id}),
+      waitForFrameNavigated(session).promise,
+    ]);
+
+    session.off('Page.backForwardCacheNotUsed', onBfCacheNotUsed);
+
+    return bfCacheEvent;
+  }
+
+  /**
+   * @param {LH.Gatherer.FRTransitionalContext<'DevtoolsLog'>} context
+   * @return {LH.Crdp.Page.BackForwardCacheNotUsedEvent[]}
+   */
+  passivelyCollectBFCacheEvent(context) {
+    const events = [];
+    for (const event of context.dependencies.DevtoolsLog) {
+      if (event.method === 'Page.backForwardCacheNotUsed') {
+        events.push(event.params);
```

Comment:
will we need the urls associated with this?

Reply:
Potentially in the future if we start surfacing more than one failure. In that scenario, it would be easy to add a url parameter to `Artifacts.BFCacheFailure`. I don't think we should worry about it right now though.